### PR TITLE
Implemented

### PR DESCRIPTION
**Pull request by Swyatik19**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,3 +7,15 @@ Make your `ArrayList` behave the same way as `java.util.ArrayList` (it should ha
...
+jv-array-list
+Створіть свій власний клас ArrayList, який реалізує інтерфейс List. 
+Перевизначте необхідні методи.
+Зробіть так, щоб ваш ArrayList поводився так само, 
+як і java.util.ArrayList (він повинен мати ємність за замовчуванням, 
+внутрішній масив повинен зростати в 1,5 рази,
+коли він заповнений, тощо).
+Викиньте користувацьке виключення ArrayListIndexOutOfBoundsException, якщо індекс, 
+переданий будь-якому з методів, є недійсним. 
+Викиньте NoSuchElementException в remove(T element), якщо такого елемента немає.
```
Comment (MaliukDaria) [inline README.md:21]: we don't need to change this file, lets remove these changes from PR

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,121 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:6]: redundant empty line

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,121 @@
...
+    private static final int DEFAULT_CAPACITY = 10;
+    private int arraySize = DEFAULT_CAPACITY;
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:8]: why do we need this variable? Can we implement solution without it?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,121 @@
...
+    private T[] dataValue;
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:9]:
```suggestion
    private T[] elements;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,121 @@
...
+    private int size;
+
+    public ArrayList() {
+        dataValue = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int newCapacity) {
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:16]:
```suggestion
    public ArrayList(int capacity) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,121 @@
...
     public void add(T value, int index) {
-
+        if (index >= 0 & index <= size) {
+            resize();
+            System.arraycopy(dataValue, index, dataValue, index + 1, size - index);
+            dataValue[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index isn't valid.");
+        }
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:36]:
```suggestion
        if (index < 0 & index > size) {
         	throw new ArrayListIndexOutOfBoundsException("Index isn't valid.");
         }
            resize();
            System.arraycopy(dataValue, index, dataValue, index + 1, size - index);
            dataValue[index] = value;
            size++;
```

it is really hard to read if-else constructions, it's better to write code without them if it possible

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,121 @@
...
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            resize();
+            dataValue[size] = list.get(i);
+            size++;
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:44]: can we use our `add(T value)` method here?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,121 @@
...
 
     @Override
     public T get(int index) {
-        return null;
+
+        if (index < size && index >= 0) {
+            return dataValue[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't exist.");
+        }
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:55]:
```suggestion
        if (index >= size && index < 0) {
        	throw new ArrayListIndexOutOfBoundsException("Index " + index + " doesn't exist sor size " + size);
        }
            return dataValue[index];
```

it's better to make exception messages more informative

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,121 @@
...
     }
 
     @Override
     public void set(T value, int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't exist.");
+        }
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:62]: It lokks like we have the same check repeated several times in code. Lets move it to separate private method

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,121 @@
...
-        return null;
+
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index doesn't exist. ");
+        }
+        final T oldValue = dataValue[index];
+        System.arraycopy(dataValue, index + 1, dataValue, index, size - index - 1);
+        dataValue[size - 1] = null;
+        size--;
+        resize();
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:77]: wy do er need to use `resize()` here?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,121 @@
...
+        int index = indexOf(element);
+        if (index >= 0) {
+            System.arraycopy(dataValue, index + 1, dataValue, index, size - index - 1);
+            dataValue[size--] = null;
+            return element;
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("There isn't such element.");
+        }
         return null;
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:92]:
```suggestion
            throw new NoSuchElementException("There isn't such element.");
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,48 +1,121 @@
...
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void resize() {
+        if (size >= arraySize) {
+            T[] newValues = (T[]) new Object[size * 3 / 2 + 1];
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:107]: lets make number 3/2 as constant field

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,48 +1,121 @@
...
+            System.arraycopy(dataValue, 0, newValues, 0, size);
+            dataValue = newValues;
+        }
+    }
+
+    public int indexOf(T valu) {
```
Comment (MaliukDaria) [inline src/main/java/core/basesyntax/ArrayList.java:113]: I think we can make this method private

**Review by MaliukDaria — CHANGES_REQUESTED** [review]

**Review by MaliukDaria — CHANGES_REQUESTED** [review]
test

**Review by Nick97-git — CHANGES_REQUESTED** [review]
you have an unresolved comments from previous review, fix them

File: src/main/java/core/basesyntax/ArrayList.java [hunk 14]
```diff
@@ -1,48 +1,96 @@
...
+
+    private void resize() {
+        if (size >= DEFAULT_CAPACITY) {
+            T[] newValues = (T[]) new Object[size * ARRAY_MULTIPLIER / DIVIDER];
+            System.arraycopy(elements, 0, newValues, 0, size);
+            elements = newValues;
+        }
+    }
+
+    private void getAudit(int index) {
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:91]:
```suggestion
    private void checkIndex(int index) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 15]
```diff
@@ -1,48 +1,96 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ARRAY_MULTIPLIER = 3;
+    private static final int DIVIDER = 2;
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:8]:
```suggestion
    private static final double GROWTH_COEFFICIENT = 1.5;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 16]
```diff
@@ -1,48 +1,96 @@
...
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void resize() {
+        if (size >= DEFAULT_CAPACITY) {
+            T[] newValues = (T[]) new Object[size * ARRAY_MULTIPLIER / DIVIDER];
```
Comment (Nick97-git) [inline src/main/java/core/basesyntax/ArrayList.java:85]:
```suggestion
            T[] newValues = (T[]) new Object[size * GROWTH_COEFFICIENT];
```

**Review by Nick97-git — CHANGES_REQUESTED** [review]

**Review by Nick97-git — APPROVED** [review]

**Pull request closed without merging**